src/error.rs

```diff
@@ -5,7 +5,7 @@ use nova::errors::NovaError;
 use thiserror::Error;
 
 #[derive(Error, Debug)]
-pub enum Error {
+pub enum ProofError {
```

Comment:
Why would we rename the most general error type from `Error` to `ProofError`? Is the idea that errors that are not somehow 'proof errors' will no longer exist?

Reply:
I named it that because it can only occur during proving. The fact that all the other errors derive from this one is probably what is incorrect here, but i wasn't ready to tackle that yet, due to it really being outside the scope of this PR. But I'm ok with trying to fix it here.

I am not sure what the hierarchy should be, maybe something like this:

* Error
  * ProofError 
    * NovaError
    * SynthesisError
  * EvalError
    * StoreError
    * ParserError

IMO what shouldn't be a thing is `LurkError`. Generally I try hard to avoid having the module or crate name as part of any type or function's name. That is a violation of modularity. You should never see `lurk::LurkError`, only `lurk::Error`. Rust provides a way to differentiate between same-named things from different modules and I try to use it whenever possible.

That would mean I would probably want `Error` to be re-exported from the top-level module so users can refer to it as `lurk::Error`.

Reply:
Okay, it's a bit hard to take all this in — but the tree you sketched above doesn't include the `Lurk` variant of `ProofError`. What's that supposed to be?

I think we're hitting a recurring friction here, which is that this is a design question that probably requires some thought and doesn't really need to be solved or even considered in order to accomplish the PR's goal. It would be better to do nothing than to propagate new structures and names that create confusion and have to be changed.

So I would rather separate this kind of thing out rather than have one-PR-to-rule-them-all which shotguns too many opinions.

Reply:
The PR description says how I arrived at making these changes - I think maybe some of the renames were ill-advised but those are trivial to change. Moving the declarations to a different module, I think was still the right thing but maybe not in this PR. Again easily reverted if you prefer. I can also just make the very minimum change of inserting `ProvableError` somewhere in the hierarchy (although I didn't think there was anywhere that made sense in the hierarchy as it was, hence my motive for the changes). But I could have just done whatever and left it to another PR to clean it up. i can still do that, if you prefer, changing this PR is not very difficult.

Reply:
In the context of my comments, I would say the simplest thing would be just creating a new error and not also significantly refactoring the errors and their hierarchy (which can be done independently, if appropriate).

Just looking back at the state of the error types *before* this PR, it seems like this could very easily have been a `LurkError`. The more energy we put into trying to create structure, the more important precision and accuracy become. I don't see what the problem with just adding this new error type would have been.

Would that work? It's extremely difficult for me to evaluate how this should work in a refactored world because it's not yet clear to me that the refactor is right, and I don't yet understand the naming/structure produced.

Again, better not to couple these things — at least without fuller understanding of how the pieces fit together.

Reply:
Here were the steps I went through (IIRC, it's been a while):

* In order to return a `Control::Return` error object in `reduce_with_witness` I'd have needed another level of nested `match` and I wanted to avoid that due to there being a lot of nesting already.
* To solve that I decided to make a separate function that returns a new Error variant that will signal that we need to return a `Control::Return`. Then I can use the ? operator and avoid the nesting.
*  In order to return the `Control::Return` it needs to know the `expr` pointer, which needed to be carried by the Error variant I created. 
* Adding that field requires a generic parameter F on the whole enum.
* Not wanting to infect LurkError with that generic param F, I moved the variant elsewhere, to the ReduceError enum.

I'm certainly willing to implement something better, if you can think of a better way to solve it. I like the idea of temporarily carrying this data as an Error, conceptually, but clearly it has tradeoffs.

Reply:
Thanks for the further context. After digging in and thinking it all through, I'm inclined to agree that the structure you ended on works for the present needs. Let's rename `ProvableError` to `ExplicitError` and call it good.

Thanks for your patience in working through the issues and meta-issues.